nixos/modules/misc/nixpkgs.nix

```diff
@@ -324,74 +324,135 @@ in
-            else config.nixpkgs.localSystem.system or (lib.systems.parse.doubleFromSystem (lib.systems.parse.mkSystemFromString config.nixpkgs.localSystem.config));
-          nixosOption =
-            if config.nixpkgs.crossSystem != null
-            then "nixpkgs.crossSystem"
-            else "nixpkgs.localSystem";
-          pkgsSystem = finalPkgs.stdenv.targetPlatform.system;
-        in {
-          assertion = constructedByMe -> !hasPlatform -> nixosExpectedSystem == pkgsSystem;
-          message = "The NixOS nixpkgs.pkgs option was set to a Nixpkgs invocation that compiles to target system ${pkgsSystem} but NixOS was configured for system ${nixosExpectedSystem} via NixOS option ${nixosOption}. The NixOS system settings must match the Nixpkgs target system.";
-        }
-      )
-      {
-        assertion = constructedByMe -> hasPlatform -> legacyOptionsDefined == [];
-        message = ''
-          Your system configures nixpkgs with the platform parameter${optionalString hasBuildPlatform "s"}:
-          ${hostPlatformLine
-          }${buildPlatformLine
-          }
-          However, it also defines the legacy options:
-          ${concatMapStrings showOptionWithDefLocs legacyOptionsDefined}
-          For a future proof system configuration, we recommend to remove
-          the legacy definitions.
+        description = mdDoc ''
+          The derivation to the nixpkgs sources, as a string with context. This
+          is set up by the nixpkgs flake.
+        '';
+      };
+
+      setNixPath = mkOption {
+        type = types.bool;
+
+        default = cfg.flake.source != null;
+        defaultText = "config.nixpkgs.flake.source != null";
+
+        description = mdDoc ''
+          Whether to create `/etc/nix/inputs/nixpkgs` as a symlink to the version of nixpkgs that
+          the NixOS system was built with and set `NIX_PATH` to include `/etc/nix/inputs/nixpkgs`.
+
+          This makes {command}`nix-build '<nixpkgs>' -A hello` work out of the box on flake systems.
         '';
-      }
-      {
-        assertion = opt.pkgs.isDefined -> cfg.config == {};
-        message = ''
-          Your system configures nixpkgs with an externally created instance.
-          `nixpkgs.config` options should be passed when creating the instance instead.
-
-          Current value:
-          ${lib.generators.toPretty { multiline = true; } opt.config}
+      };
+
+      setFlakeRegistry = mkOption {
+        type = types.bool;
+
+        default = cfg.flake.source != null;
+        defaultText = "config.nixpkgs.flake.source != null";
+
+        description = mdDoc ''
+          Whether to pin nixpkgs in the system-wide flake registry (`/etc/nix/registry.json`) to the
+          version of nixpkgs that the NixOS system has been built with.
+
+          This also sets a default to disable the global (online) flake registry due to
+          <https://github.com/NixOS/nix/issues/9087>.
```

Comment:
ah good call

Reply:
Bad call: unnecessary breaking change. Why not fix the bug instead?

Reply:
I don't want to try getting anything into Nix after a one line change to fix nix-shell in shebangs by a former coworker continues to sit unreviewed for going on two years now, in spite of repeated prodding for reviews including by back channels.

I understand the Nix team has been improving things, and I think more prompt reviews generally happen now but I'm still not interested.

I can remove the workaround because the bug isn't that bad.

Reply:
@lf- i have an interest in the shebangs features, but I am not finding the referenced shebang PR. (my own effort in bringing shebangs to nix-command took more than two years with 5189, so I understand the patience can get worn out)

Reply:
i had it written down somewhere; it's https://github.com/NixOS/nix/pull/5088